src/vs/editor/contrib/hover/browser/contentHover.ts

```diff
@@ -437,252 +442,385 @@ class ContentHoverVisibleData {
+		}
+		this._setHoverData(hoverData);
+		this._updateFont();
+		this._updateContent(node);
+		this._updateContentsDomNodeMaxDimensions();
+		this.onContentsChanged();
 		// Simply force a synchronous render on the editor
 		// such that the widget does not really render with left = '0px'
 		this._editor.render();
+	}
+
+	private _setContentPosition(hoverData: HoverData, preference?: ContentWidgetPositionPreference) {
+		this._contentPosition = {
+			position: hoverData.showAtPosition,
+			secondaryPosition: hoverData.showAtSecondaryPosition,
+			positionAffinity: hoverData.isBeforeContent ? PositionAffinity.LeftOfInjectedText : undefined,
+			preference: [preference ?? ContentWidgetPositionPreference.ABOVE]
+		};
+	}
+
+	public showAt(node: DocumentFragment, hoverData: HoverData): void {
+		if (!this._editor || !this._editor.hasModel()) {
+			return;
+		}
+		this._setContentPosition(hoverData);
+		this._render(node, hoverData);
+		const widgetHeight = this._getWidgetHeight();
+		const widgetPosition = hoverData.showAtPosition;
+		this._positionPreference = this._findPositionPreference(widgetHeight, widgetPosition) ?? ContentWidgetPositionPreference.ABOVE;
+		this._setContentPosition(hoverData, this._positionPreference);
 
 		// See https://github.com/microsoft/vscode/issues/140339
 		// TODO: Doing a second layout of the hover after force rendering the editor
 		this.onContentsChanged();
-
-		if (visibleData.stoleFocus) {
-			this._hover.containerDomNode.focus();
+		if (hoverData.stoleFocus) {
+			this._hoverWidget.containerDomNode.focus();
 		}
-		visibleData.colorPicker?.layout();
+		hoverData.colorPicker?.layout();
 	}
 
 	public hide(): void {
-		if (this._visibleData) {
-			const stoleFocus = this._visibleData.stoleFocus;
-			this._setVisibleData(null);
-			this._hoverFocusedKey.set(false);
-			this._editor.layoutContentWidget(this);
-			if (stoleFocus) {
-				this._editor.focus();
-			}
+		if (!this._hoverData) {
+			return;
 		}
-	}
-
-	public onContentsChanged(): void {
+		this._setHoverData(undefined);
+		this._resizableNode.maxSize = new dom.Dimension(Infinity, Infinity);
+		this._resizableNode.clearSashHoverState();
+		this._editor.removeContentWidget(this);
+		this._hoverFocusedKey.set(false);
 		this._editor.layoutContentWidget(this);
-		this._hover.onContentsChanged();
-
-		const scrollDimensions = this._hover.scrollbar.getScrollDimensions();
-		const hasHorizontalScrollbar = (scrollDimensions.scrollWidth > scrollDimensions.width);
-		if (hasHorizontalScrollbar) {
-			// There is just a horizontal scrollbar
-			const extraBottomPadding = `${this._hover.scrollbar.options.horizontalScrollbarSize}px`;
-			if (this._hover.contentsDomNode.style.paddingBottom !== extraBottomPadding) {
-				this._hover.contentsDomNode.style.paddingBottom = extraBottomPadding;
-				this._editor.layoutContentWidget(this);
-				this._hover.onContentsChanged();
-			}
+		if (this._hoverData.stoleFocus) {
+			this._editor.focus();
```

Comment:
While testing your changes I saw this erroring (see below). That's because of set `_setHoverData` being called a few lines above

```ERR Cannot read properties of undefined (reading 'stoleFocus'): TypeError: Cannot read properties of undefined (reading 'stoleFocus')
    at ResizableHoverWidget.hide (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/contentHover.js:602:33)
    at ContentHoverController._setCurrentResult (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/contentHover.js:161:30)
    at ContentHoverController.hide (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/contentHover.js:167:18)
    at ModesHoverController._hideWidgets (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/hover.js:162:38)
    at ModesHoverController._onEditorMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/hover.js:94:18)
    at vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/hover.js:56:70
    at Listener.invoke (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/base/common/event.js:660:27)
    at EventDeliveryQueue.deliver (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/base/common/event.js:836:38)
    at InteractionEmitter.fire (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/base/common/event.js:800:37)
    at InteractionEmitter.fire (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/widget/codeEditorWidget.js:1431:19)
    at viewUserInputEvents.onMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/widget/codeEditorWidget.js:1318:74)
    at ViewUserInputEvents.emitMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/view/viewUserInputEvents.js:37:32)
    at ViewController.emitMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/view/viewController.js:248:34)
    at TouchHandler._onMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/controller/mouseHandler.js:193:33)
    at vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/controller/mouseHandler.js:40:94
    at HTMLDivElement.<anonymous> (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/editorDom.js:131:17)
```